pelican/plugins/pelican_redirect/__init__.py

```diff
@@ -35,28 +43,39 @@ def __init__(self, *args, **kwargs):
 
     def redirect_articles(self, article_generator):
         for article in article_generator.articles:
-            if article.metadata.get("original_url"):
-                self.redirects.append(article)
+            for url in self.historical_urls(article_generator, article, kind="ARTICLE"):
+                self.redirects.append(Redirect(to=article, from_url=url))
 
     def redirect_pages(self, page_generator):
         for page in page_generator.pages:
-            if page.metadata.get("original_url"):
-                self.redirects.append(page)
+            for url in self.historical_urls(page_generator, page, kind="PAGE"):
+                self.redirects.append(Redirect(to=page, from_url=url))
+
+    def historical_urls(self, generator, page, kind):
+        for formats in generator.settings.get("PAST_ARTICLE_URL_FORMATS", {}).values():
+            if f"{kind}_URL" in formats:
+                yield self.format_url_from_page(page, formats[f"{kind}_URL"])
+
+        if page.metadata.get("original_url"):
+            yield page.metadata.get("original_url")
+
+    def format_url_from_page(self, page, url_format):
```

Comment:
It is, yeah, but if Pelican changes/alters how it generates URLs that might change, and also there are some interesting bits in how the url format string is chosen in the Pelican code that I'm only _mostly_ sure that I understood, so I figured to wrap this up in case we wanted something more complex later (it's also much more amenable to unit testing)